### ISSUE TITLE
MarkWritable

### ISSUE THREAD

**Issue by IlyaSemenov** [ts-essentials/ts-essentials#294]
Would it be possible to implement `MarkWritable` similar to how `MarkRequired` works?

The use is case is when there is a public context interface where a field is marked as readonly:

```ts
interface RequestContext {
  readonly pluginHelper: PluginHelper // third party code should not be able to modify this field
}
```

however, the actual plugin must be able to actually install this helper:

```ts
export function pluginMiddleware(ctx: RequestContext) {
  const writableCtx = MarkWritable<ctx, "pluginHelper"> // allow assignment to read-only property
  writableCtx.pluginHelper = new PluginHelper()
}
```

I tried a naive rework of `MarkRequired`:

```ts
import { Writable } from "ts-essentials"

export declare type MarkWritable<T, WK extends keyof T> = Exclude<T, WK> &
	Writable<Pick<T, WK>>
```

however it doesn't work (still results in *Cannot assign to 'foo' because it is a read-only property.ts(2540)*)

**Comment by Beraliv** [comment]
Hey, @IlyaSemenov 👋

Thank you for the report and the suggestion! 🚀

I checked the example with simple object (`string` instead of `PluginHelper`) in Playground and it's working fine – https://tsplay.dev/NaYaBN

What TypeScript version do you use? What is in your `tsconfig.json`? Can you please try to reproduce it in Playground somehow and share it with me? That would be helpful

Also you can try to put `PluginHelper` into the Playground to be able to be as close to the real example as possible

Thank you in advance!

And of course, I don't mind adding `MarkWritable` to our codebase!

I will add it on the weekend and if we're able to reproduce it together, we will add more test cases or fix it on the project side.

**Comment by IlyaSemenov** [comment]
@Beraliv I traced the problem to the fact that you've changed `Exclude<T,K>` to `Omit<T,K>` in your tsplay fiddle. That is a valid change which makes things work, I overlooked that indeed.

With that in mind, I now wonder why `MarkRequired` is using `Exclude` and not `Omit`?

https://github.com/krzkaczor/ts-essentials/blob/61f13d12c69d5a2f4d9ece9cd24b9b56c20c2936/lib/types.ts#L392

This seems to be a mistake.

**Comment by Beraliv** [comment]
Hey again! 👋

I'm coming back to you with the changes that we recently discussed:

1. I've fixed `MarkRequired` – https://github.com/krzkaczor/ts-essentials/pull/295 which only improves DX
2. I've added `MarkWritable` – https://github.com/krzkaczor/ts-essentials/pull/296

I will try to merge it and release within a week

I will keep you updated here

**Comment by IlyaSemenov** [comment]
I think similarly there should be a counterpart to MarkWritable named MarkReadonly. Not that I have a particular use case for that, but just for consistency and symmetry as it goes everywhere else.

**Comment by Beraliv** [comment]
> I think similarly there should be a counterpart to MarkWritable named MarkReadonly. Not that I have a particular use case for that, but just for consistency and symmetry as it goes everywhere else.

That's a good suggestion!

I will add it as well

**Comment by Beraliv** [comment]
Done here – https://github.com/krzkaczor/ts-essentials/pull/298 ✅

I will let you know when everything's merged and released

**Comment by Beraliv** [comment]
It will be released under https://github.com/ts-essentials/ts-essentials/pull/302

**Comment by Beraliv** [comment]
Leaving comment here as I didn't release it yet.

Please ping me if it's urgent, otherwise I will release it in the beginning of May when my schedule comes to normal.

Sorry for the inconvenience!

**Comment by IlyaSemenov** [comment]
That's fine, take your time!

**Comment by Beraliv** [comment]
Released it in https://github.com/ts-essentials/ts-essentials/pull/302 and it's available now in 9.2.0 – https://github.com/ts-essentials/ts-essentials/releases/tag/v9.2.0
